### PR TITLE
Adding support for ComputeCpp compiler

### PR DESCRIPTION
**Pull request by joeatodd**
CMake changes, minor code changes & notes for ComputeCpp support.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,17 +1,25 @@
...
 Forked from https://github.com/salel/nbody
 
-## Building
+## Compilers/Backends
+
+This nbody simulation can be run with any of:
+ - CUDA
+ - DPC++ CUDA backend
+ - DPC++ OpenCL CPU backend
+ - ComputeCpp Experimental Host Backend
```
Comment (DuncanMcBain) [inline README.md]:
```suggestion
 - ComputeCpp OpenCL Backend
```

File: README.md [hunk 2]
```diff
@@ -21,37 +29,54 @@ These can be installed with apt:
...
-> sycl-ls
-[opencl:cpu:0] Intel(R) OpenCL, Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz 3.0 [2021.13.11.0.23_160000]
-[opencl:cpu:1] Intel(R) OpenCL, Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz 3.0 [2021.13.11.0.23_160000]
-[cuda:gpu:0] NVIDIA CUDA BACKEND, NVIDIA GeForce RTX 3060 0.0 [CUDA 11.6]
-[host:host:0] SYCL host platform, SYCL host device 1.2 [1.2]
-```
+Both DPC++ backends require the [DPC++ compiler](https://intel.github.io/llvm-docs/GetStartedGuide.html) to compile the SYCL code.
 
-### Converting CUDA to SYCL
+The ComputeCpp Experimental Host backend requires the proprietary ComputeCpp binary which you should have received from Codeplay, and an OpenCL runtime.
```
Comment (DuncanMcBain) [inline README.md]:
```suggestion
The ComputeCpp backend requires ComputeCpp which you can download from the [Codeplay website](https://developer.codeplay.com/home/).
```

File: README.md [hunk 3]
```diff
@@ -60,13 +85,21 @@ will run on the CUDA backend, whereas:
...
 
-**Note**: this is possible because `CMakeLists.txt` specifies building the SYCL code for both CUDA & OpenCL backends:
+**Note**: Selection between DPC++ backends at runtime is possible because `CMakeLists.txt` specifies building the SYCL code for both CUDA (`nvptx64-nvidia-cuda`) & OpenCL (`spir64`) targets:
 ```
      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsycl -fsycl-targets=spir64,nvptx64-nvidia-cuda -fsycl-unnamed-lambda")
 ```
 
+### Selecting the ComputeCpp Host Backend
+
+The experimental ComputeCpp distribution will *only* work with the host backend, and this must be specified with the `COMPUTECPP_TARGET` environment variable:
```
Comment (DuncanMcBain) [inline README.md]: Not 100% sure what to say here but if the user downloads ComputeCpp, it should work with either the host device or any USM-supporting OpenCL device (e.g. Intel GPU/CPU).

File: cmake/FindComputeCpp.cmake [hunk 4]
```diff
@@ -0,0 +1,561 @@
...
+#
+#  Tools for finding and building with ComputeCpp.
+#
+#  User must define ComputeCpp_DIR pointing to the ComputeCpp
+#  installation.
+#
+#  Latest version of this file can be found at:
+#    https://github.com/codeplaysoftware/computecpp-sdk
+
+cmake_minimum_required(VERSION 3.4.3)
```
Comment (DuncanMcBain) [inline cmake/FindComputeCpp.cmake]: So just to be annoying, I recently updated the upstream version of this file. It's fine to copy it but there are improvements and bugfixes upstream that IMO would help in this PR.

File: scripts/build_computecpp.sh [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+rm -rf $BUILD_DIR
+mkdir $BUILD_DIR
+cd $BUILD_DIR || exit
+
+cmake ../ \
+	-DGLEW_LIBRARY=/usr/lib/x86_64-linux-gnu/libGLEW.so \
+	-DBUILD_SYCL=on \
+	-DUSE_COMPUTECPP=on \
+	-DComputeCpp_DIR=/home/joetodd/sources/ComputeCpp \
+	-DCOMPUTECPP_BITCODE="host-x86_64" \
```
Comment (DuncanMcBain) [inline scripts/build_computecpp.sh]: I'd probably remove this since it's going to strange in most cases.

File: scripts/build_computecpp.sh [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+	-DSYCL_LANGUAGE_VERSION=202002\
```
Comment (DuncanMcBain) [inline scripts/build_computecpp.sh]:
```suggestion
	-DSYCL_LANGUAGE_VERSION=2020\
```
There are newer versions than this, setting it to 2020 will get the newest of these IIRC

File: scripts/build_computecpp.sh [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+	-DOpenCL_LIBRARY=/usr/lib/x86_64-linux-gnu/libOpenCL.so.1 \
```
Comment (DuncanMcBain) [inline scripts/build_computecpp.sh]: If this needs to be here to work on your machine, fair enough, but on machines set up for ComputeCpp ideally the system installs of each should be sufficient (e.g. `/usr/lib`, `/usr/include`).

File: src_sycl/CMakeLists.txt [hunk 8]
```diff
@@ -21,14 +21,43 @@ add_executable(nbodygl ${COMMON_SOURCE} ${OPENGL_SOURCE})
...
 
-if(CMAKE_CXX_COMPILER MATCHES ".*dpcpp[.exe]*")
-	target_compile_options(nbodygl PRIVATE -fsycl)
-endif()
+option(USE_COMPUTECPP "whether to use ComputeCpp compiler" OFF)
+
+if(USE_COMPUTECPP)
+	find_package(ComputeCpp REQUIRED)
+
+	set(CMAKE_CXX_COMPILER ${ComputeCpp_DEVICE_COMPILER_EXECUTABLE})
```
Comment (DuncanMcBain) [inline src_sycl/CMakeLists.txt:35]: This isn't necessary, but is possible. If you do do it, then it sets `-sycl-driver` mode automatically in `FindComputeCpp.cmake`.

File: src_sycl/CMakeLists.txt [hunk 9]
```diff
@@ -21,14 +21,43 @@ add_executable(nbodygl ${COMMON_SOURCE} ${OPENGL_SOURCE})
...
+
+	set_property(SOURCE simulator.dp.cpp APPEND
+        PROPERTY COMPILE_OPTIONS
+		-sycl-driver -std=gnu++17 -O3
+		-mllvm -local_mem_pools=8
+		-mllvm -local_mem_pools_size=5000
+		-march=native -sycl-target host-x86_64)
```
Comment (DuncanMcBain) [inline src_sycl/CMakeLists.txt]: I don't know that any of these should be necessary. `-sycl-driver` can be implied, `-O3` is the default, new versions of FindComputeCpp.cmake will set `-std=xx` based on the `CXX_STANDARD` CMake property is, and the other output stuff is set by the `COMPUTECPP_BITCODE` options and so on.

If you don't use the sycl driver mode then you could set `-march=native` on the host compiler instead, to get some benefits to the generated host code.

File: src_sycl/CMakeLists.txt [hunk 10]
```diff
@@ -21,14 +21,43 @@ add_executable(nbodygl ${COMMON_SOURCE} ${OPENGL_SOURCE})
...
+
+	target_include_directories(nbodygl PRIVATE
+		${ComputeCpp_INCLUDE_DIRS} ${OpenCL_INCLUDE_DIR})
 
+	target_link_libraries(nbodygl PRIVATE ComputeCpp::ComputeCpp)
 
+	target_compile_definitions(nbodygl INTERFACE
+    	SYCL_LANGUAGE_VERSION=${SYCL_LANGUAGE_VERSION})
```
Comment (DuncanMcBain) [inline src_sycl/CMakeLists.txt]: These three should all be set by FindComputeCpp.cmake. If they're not I'd consider that a serious bug and promise I'll fix it!

File: src_sycl/CMakeLists.txt (hunk 9 above)
Comment (joeatodd) [reply src_sycl/CMakeLists.txt]: Found that, with the current setup, `-O3` does make a difference, but have removed the rest.

File: src_sycl/CMakeLists.txt (hunk 10 above)
Comment (joeatodd) [reply src_sycl/CMakeLists.txt]: They were!

**Pull request merged**